drivers/led/lp5562.c

```diff
@@ -0,0 +1,837 @@
+#define LP5562_PROG_COMMAND_RAMP_TIME(prescale, step_time) \
+	(((prescale) << 6) | (step_time))
+#define LP5562_PROG_COMMAND_STEP_COUNT(fade_direction, count) \
+	(((fade_direction) << 7) | (count))
+
+/* Helper definitions. */
+#define LP5562_PROG_MAX_COMMANDS 16
+#define LP5562_MASK              0x03
+#define LP5562_CHANNEL_MASK(channel) ((LP5562_MASK) << (channel << 1))
+
+/*
+ * Available channels. There are four LED channels usable with the LP5562. While
+ * they can be mapped to LEDs of any color, the driver's typical application is
+ * with a red, a green, a blue and a white LED. Since the data sheet's
+ * nomenclature uses RGBW, we keep it that way.
+ */
+enum lp5562_led_channels {
+	LP5562_CHANNEL_B,
+	LP5562_CHANNEL_G,
+	LP5562_CHANNEL_R,
+	LP5562_CHANNEL_W,
+
+	LP5562_CHANNEL_COUNT,
+};
+
+/*
+ * Each channel can be driven by directly assigning a value between 0 and 255 to
+ * it to drive the PWM or by one of the three execution engines that can be
+ * programmed for custom lighting patterns in order to reduce the I2C traffic
+ * for repetitive patterns.
+ */
+enum lp5562_led_sources {
+	LP5562_SOURCE_PWM,
+	LP5562_SOURCE_ENGINE_1,
+	LP5562_SOURCE_ENGINE_2,
+	LP5562_SOURCE_ENGINE_3,
+
+	LP5562_SOURCE_COUNT,
+};
+
+/* Operational modes of the execution engines. */
+enum lp5562_engine_op_modes {
+	LP5562_OP_MODE_DISABLED = 0x00,
+	LP5562_OP_MODE_LOAD = 0x01,
+	LP5562_OP_MODE_RUN = 0x02,
+	LP5562_OP_MODE_DIRECT_CTRL = 0x03,
+};
+
+/* Execution state of the engines. */
+enum lp5562_engine_exec_states {
+	LP5562_ENGINE_MODE_HOLD = 0x00,
+	LP5562_ENGINE_MODE_STEP = 0x01,
+	LP5562_ENGINE_MODE_RUN = 0x02,
+	LP5562_ENGINE_MODE_EXEC = 0x03,
+};
+
+/* Fading directions for programs executed by the engines. */
+enum lp5562_engine_fade_dirs {
+	LP5562_FADE_UP = 0x00,
+	LP5562_FADE_DOWN = 0x01,
+};
+
+struct lp5562_data {
+	struct device *i2c;
+	struct led_data dev_data;
+};
+
+/*
+ * @brief Get the register for the given LED channel used to directly write a
+ *	brightness value instead of using the execution engines.
+ *
+ * @param channel LED channel.
+ * @param source Buffer for the register.
+ *
+ * @return 0 if successful, -ERRNO otherwise.
+ */
+static int lp5562_get_pwm_reg(enum lp5562_led_channels channel, u8_t *reg)
+{
+	switch (channel) {
+	case LP5562_CHANNEL_W:
+		*reg = LP5562_W_PWM;
+		break;
+	case LP5562_CHANNEL_R:
+		*reg = LP5562_R_PWM;
+		break;
+	case LP5562_CHANNEL_G:
+		*reg = LP5562_G_PWM;
+		break;
+	case LP5562_CHANNEL_B:
+		*reg = LP5562_B_PWM;
+		break;
+	default:
+		SYS_LOG_ERR("Invalid channel given.");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * @brief Get the base address for programs of the given execution engine.
+ *
+ * @param engine Engine the base address is requested for.
+ * @param base_addr Buffer for the base address.
+ *
+ * @return 0 if successful, -ERRNO otherwise.
+ */
+static int lp5562_get_engine_ram_base_addr(enum lp5562_led_sources engine,
+					   u8_t *base_addr)
```

Comment:
Please fix up the alignment here

Reply:
Can you check in an editor? In vim it looks like the alignment is right, so maybe it is just the GitHub rendering.

Reply:
I was referrig to the mis alignment regarding the second parameter `u8_t *base_addr`. It should have started at the level after opening parentheses (i.e) below `e`. Btw, this is just a nitpick which will make your code look good :)

Reply:
sure, that's what I understood and I actually did align those arguments. What I see locally is different to what I see in the GitHub rendering :man_shrugging: 
 
![lp5562](https://user-images.githubusercontent.com/18048777/43153736-67e6a4b2-8f72-11e8-9089-d83a55f2e27c.jpg)

But then again I am relying on a 8 space width of tabs for that of course. But this is what the other drivers look like as well